File: android/app/src/main/java/beaver/notes/pocket/WebDAVPlugin.java
package beaver.notes.pocket;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import android.util.Base64;
import android.util.Log;

import java.io.IOException;
import java.security.KeyStore;
import java.security.cert.Certificate;
import java.security.cert.CertificateFactory;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.*;

import okhttp3.*;

@CapacitorPlugin(name = "WebDAV")
public class WebDAVPlugin extends Plugin {

    private OkHttpClient client;
    private boolean useInsecure = false;
    private SSLSocketFactory sslSocketFactory;
    private X509TrustManager trustManager;

    @Override
    public void load() {
        rebuildClient();
    }

    // Rebuild OkHttpClient depending on secure/insecure mode
    private void rebuildClient() {
        try {
            if (useInsecure) {
                TrustManager[] trustAllCerts = new TrustManager[] {
                        new X509TrustManager() {
                            public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) {}
                            public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) {}
                            public java.security.cert.X509Certificate[] getAcceptedIssuers() { return new java.security.cert.X509Certificate[]{}; }
                        }
                };

                SSLContext sslContext = SSLContext.getInstance("SSL");
                sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
                sslSocketFactory = sslContext.getSocketFactory();

                trustManager = (X509TrustManager) trustAllCerts[0];

                client = new OkHttpClient.Builder()
                        .sslSocketFactory(sslSocketFactory, trustManager)
                        .hostnameVerifier((hostname, session) -> true)
                        .connectTimeout(30, TimeUnit.SECONDS)
                        .readTimeout(30, TimeUnit.SECONDS)
                        .build();
            } else if (sslSocketFactory != null && trustManager != null) {
                // Custom cert uploaded (sslSocketFactory & trustManager set)
                client = new OkHttpClient.Builder()
                        .sslSocketFactory(sslSocketFactory, trustManager)
                        .hostnameVerifier((hostname, session) -> true)
                        .connectTimeout(30, TimeUnit.SECONDS)
                        .readTimeout(30, TimeUnit.SECONDS)
                        .build();
            } else {
                // Default OkHttpClient
                client = new OkHttpClient.Builder()
                        .connectTimeout(30, TimeUnit.SECONDS)
                        .readTimeout(30, TimeUnit.SECONDS)
                        .build();
            }
        } catch (Exception e) {
            Log.e("WebDAVPlugin", "Error rebuilding client: " + e.getMessage());
            client = new OkHttpClient();
        }
    }

    private String getAuthHeader(String username, String password) {
        String credential = username + ":" + password;
        return "Basic " + Base64.encodeToString(credential.getBytes(), Base64.NO_WRAP);
    }

    @PluginMethod
    public void setInsecureMode(PluginCall call) {
        Boolean insecure = call.getBoolean("insecure");
        if (insecure == null) {
            call.reject("Missing 'insecure' parameter");
            return;
        }
        this.useInsecure = insecure;
        rebuildClient();
        JSObject ret = new JSObject();
        ret.put("message", "Insecure mode set to " + useInsecure);
        call.resolve(ret);
    }

    @PluginMethod
    public void uploadCertificate(PluginCall call) {
        String base64Cert = call.getString("certificate");
        if (base64Cert == null) {
            call.reject("Missing 'certificate' parameter");
            return;
        }
        try {
            byte[] certBytes = Base64.decode(base64Cert, Base64.DEFAULT);
            CertificateFactory cf = CertificateFactory.getInstance("X.509");
            Certificate cert = cf.generateCertificate(new java.io.ByteArrayInputStream(certBytes));

            KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
            ks.load(null);
            ks.setCertificateEntry("ca", cert);

            TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
            tmf.init(ks);

            TrustManager[] trustManagers = tmf.getTrustManagers();
            if (trustManagers.length != 1 || !(trustManagers[0] instanceof X509TrustManager)) {
                call.reject("Unexpected default trust managers");
                return;
            }

            trustManager = (X509TrustManager) trustManagers[0];

            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, new TrustManager[]{trustManager}, null);
            sslSocketFactory = sslContext.getSocketFactory();

            rebuildClient();

            JSObject ret = new JSObject();
            ret.put("message", "Certificate uploaded successfully");
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Invalid certificate data: " + e.getMessage());
        }
    }

    @PluginMethod
    public void createFolder(PluginCall call) {
        String url = call.getString("url");
        String username = call.getString("username");
        String password = call.getString("password");

        if (url == null || username == null || password == null) {
            call.reject("Missing parameters");
            return;
        }

        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", getAuthHeader(username, password))
                .method("MKCOL", null)
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call callInner, IOException e) {
                call.reject("Failed to create folder: " + e.getMessage(), e);
            }

            @Override
            public void onResponse(Call callInner, Response response) throws IOException {
                if (response.isSuccessful()) {
                    JSObject result = new JSObject();
                    result.put("message", "Folder created successfully.");
                    call.resolve(result);
                } else {
                    String body = response.body() != null ? response.body().string() : "No response body";
                    call.reject("Failed to create folder: " + response.code() + " - " + body);
                }
            }
        });
    }

    @PluginMethod
    public void checkFolderExists(PluginCall call) {
        String url = call.getString("url");
        String username = call.getString("username");
        String password = call.getString("password");

        if (url == null || username == null || password == null) {
            call.reject("Missing parameters");
            return;
        }

        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", getAuthHeader(username, password))
                .method("PROPFIND", null)
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call callInner, IOException e) {
                call.reject("Failed to check folder existence: " + e.getMessage(), e);
            }

            @Override
            public void onResponse(Call callInner, Response response) throws IOException {
                if (response.isSuccessful()) {
                    JSObject result = new JSObject();
                    result.put("message", "Folder exists.");
                    result.put("exists", true);
                    call.resolve(result);
                } else {
                    String body = response.body() != null ? response.body().string() : "No response body";
                    call.reject("Folder does not exist: " + response.code() + " - " + body);
                }
            }
        });
    }

    @PluginMethod
    public void listContents(PluginCall call) {
        String url = call.getString("url");
        String username = call.getString("username");
        String password = call.getString("password");

        if (url == null || username == null || password == null) {
            call.reject("Missing parameters");
            return;
        }

        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", getAuthHeader(username, password))
                .method("PROPFIND", null)
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call callInner, IOException e) {
                call.reject("Failed to list folder contents: " + e.getMessage(), e);
            }

            @Override
            public void onResponse(Call callInner, Response response) throws IOException {
                if (response.isSuccessful()) {
                    String xmlResponse = response.body() != null ? response.body().string() : "No response body";
                    JSObject result = new JSObject();
                    result.put("message", "Folder contents retrieved.");
                    result.put("data", xmlResponse);
                    call.resolve(result);
                } else {
                    String body = response.body() != null ? response.body().string() : "No response body";
                    call.reject("Failed to list folder contents: " + response.code() + " - " + body);
                }
            }
        });
    }

    @PluginMethod
    public void uploadFile(PluginCall call) {
        String url = call.getString("url");
        String username = call.getString("username");
        String password = call.getString("password");
        String base64Content = call.getString("content");

        if (url == null || username == null || password == null || base64Content == null) {
            call.reject("Missing parameters");
            return;
        }

        byte[] data = Base64.decode(base64Content, Base64.DEFAULT);

        RequestBody requestBody = RequestBody.create(data);

        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", getAuthHeader(username, password))
                .put(requestBody)
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call callInner, IOException e) {
                call.reject("Upload failed: " + e.getMessage(), e);
            }

            @Override
            public void onResponse(Call callInner, Response response) throws IOException {
                if (response.isSuccessful()) {
                    JSObject result = new JSObject();
                    result.put("message", "File uploaded");
                    call.resolve(result);
                } else {
                    String body = response.body() != null ? response.body().string() : "No response body";
                    call.reject("Upload failed with status: " + response.code() + " - " + body);
                }
            }
        });
    }

    @PluginMethod
    public void deleteFolder(PluginCall call) {
        String url = call.getString("url");
        String username = call.getString("username");
        String password = call.getString("password");

        if (url == null || username == null || password == null) {
            call.reject("Missing parameters");
            return;
        }

        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", getAuthHeader(username, password))
                .delete()
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call callInner, IOException e) {
                call.reject("Delete failed: " + e.getMessage(), e);
            }

            @Override
            public void onResponse(Call callInner, Response response) throws IOException {
                if (response.isSuccessful()) {
                    JSObject result = new JSObject();
                    result.put("message", "Folder deleted");
                    call.resolve(result);
                } else {
                    String body = response.body() != null ? response.body().string() : "No response body";
                    call.reject("Delete failed with status: " + response.code() + " - " + body);
                }
            }
        });
    }

    @PluginMethod
    public void getFile(PluginCall call) {
        String url = call.getString("url");
        String username = call.getString("username");
        String password = call.getString("password");

        if (url == null || username == null || password == null) {
            call.reject("Missing parameters");
            return;
        }

        Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", getAuthHeader(username, password))
                .get()
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call callInner, IOException e) {
                call.reject("GET failed: " + e.getMessage(), e);
            }

            @Override
            public void onResponse(Call callInner, Response response) throws IOException {
                if (response.isSuccessful()) {
                    byte[] bytes = response.body() != null ? response.body().bytes() : null;
                    if (bytes != null) {
                        String base64Data = Base64.encodeToString(bytes, Base64.NO_WRAP);
                        JSObject result = new JSObject();
                        result.put("content", base64Data);
                        call.resolve(result);
                    } else {
                        call.reject("Empty response body");
                    }
                } else {
                    String body = response.body() != null ? response.body().string() : "No response body";
                    call.reject("GET failed with status: " + response.code() + " - " + body);
                }
            }
        });
    }
}

File: ios/App/App/WebDAVPlugin.swift
import Foundation
import Capacitor

@objc(WebDAVPlugin)
public class WebDAVPlugin: CAPPlugin, CAPBridgedPlugin {
     
     public override init() {
         super.init()
     }
     
     public let identifier = "WebDAVPlugin"
     public let jsName = "WebDAV"
     
     public let pluginMethods: [CAPPluginMethod] = [
          CAPPluginMethod(name: "setInsecureMode", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "uploadCertificate", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "createFolder", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "checkFolderExists", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "listContents", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "uploadFile", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "deleteFolder", returnType: CAPPluginReturnPromise),
          CAPPluginMethod(name: "getFile", returnType: CAPPluginReturnPromise)
     ]
     
     private var useInsecure: Bool = false
     private var trustedCertificate: SecCertificate?
     
     override public func load() {}
     
     @objc func setInsecureMode(_ call: CAPPluginCall) {
          useInsecure = call.getBool("insecure") ?? false
          call.resolve(["message": "Insecure mode set to \(useInsecure)"])
     }
     
     @objc func uploadCertificate(_ call: CAPPluginCall) {
          guard let base64Cert = call.getString("certificate"),
                let certData = Data(base64Encoded: base64Cert),
                let certificate = SecCertificateCreateWithData(nil, certData as CFData) else {
               call.reject("Invalid certificate data")
               return
          }
          
          self.trustedCertificate = certificate
          call.resolve(["message": "Certificate uploaded successfully"])
     }
     
     private class InsecureURLSessionDelegate: NSObject, URLSessionDelegate {
          func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                          completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
               if let serverTrust = challenge.protectionSpace.serverTrust {
                    completionHandler(.useCredential, URLCredential(trust: serverTrust))
               } else {
                    completionHandler(.performDefaultHandling, nil)
               }
          }
     }
     
     private class CustomCertSessionDelegate: NSObject, URLSessionDelegate {
          let trustedCert: SecCertificate?
          
          init(cert: SecCertificate?) {
               self.trustedCert = cert
          }
          
          func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                          completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
               guard let serverTrust = challenge.protectionSpace.serverTrust,
                     let trustedCert = trustedCert else {
                    completionHandler(.performDefaultHandling, nil)
                    return
               }
               
               let certCount = SecTrustGetCertificateCount(serverTrust)
               if certCount > 0,
                  let serverCertificate = SecTrustGetCertificateAtIndex(serverTrust, 0) {
                    let serverCertData = SecCertificateCopyData(serverCertificate) as Data
                    let localCertData = SecCertificateCopyData(trustedCert) as Data
                    
                    if serverCertData == localCertData {
                         completionHandler(.useCredential, URLCredential(trust: serverTrust))
                         return
                    }
               }
               
               completionHandler(.cancelAuthenticationChallenge, nil)
          }
     }
     
     private func getSession() -> URLSession {
          if useInsecure {
               let config = URLSessionConfiguration.default
               return URLSession(configuration: config, delegate: InsecureURLSessionDelegate(), delegateQueue: nil)
          } else if let cert = trustedCertificate {
               let config = URLSessionConfiguration.default
               return URLSession(configuration: config, delegate: CustomCertSessionDelegate(cert: cert), delegateQueue: nil)
          } else {
               return URLSession.shared
          }
     }
     
     private func getAuthHeader(username: String, password: String) -> String {
          let credentialData = "\(username):\(password)".data(using: .utf8)!
          let base64Credentials = credentialData.base64EncodedString()
          return "Basic \(base64Credentials)"
     }
     
     @objc func createFolder(_ call: CAPPluginCall) {
          guard let url = call.getString("url"),
                let username = call.getString("username"),
                let password = call.getString("password"),
                let requestURL = URL(string: url) else {
               call.reject("Missing or invalid parameters")
               return
          }
          
          var request = URLRequest(url: requestURL)
          request.httpMethod = "MKCOL"
          request.addValue(getAuthHeader(username: username, password: password), forHTTPHeaderField: "Authorization")
          
          let task = getSession().dataTask(with: request) { _, response, error in
               DispatchQueue.main.async {
                    if let error = error {
                         call.reject("Failed to create folder: \(error.localizedDescription)")
                         return
                    }
                    
                    guard let httpResponse = response as? HTTPURLResponse else {
                         call.reject("Invalid response")
                         return
                    }
                    
                    if (200...299).contains(httpResponse.statusCode) {
                         call.resolve(["message": "Folder created successfully."])
                    } else {
                         call.reject("Failed to create folder: \(httpResponse.statusCode)")
                    }
               }
          }
          task.resume()
     }
     
     @objc func checkFolderExists(_ call: CAPPluginCall) {
          guard let url = call.getString("url"),
                let username = call.getString("username"),
                let password = call.getString("password"),
                let requestURL = URL(string: url) else {
               call.reject("Missing or invalid parameters")
               return
          }
          
          var request = URLRequest(url: requestURL)
          request.httpMethod = "PROPFIND"
          request.addValue(getAuthHeader(username: username, password: password), forHTTPHeaderField: "Authorization")
          
          let task = getSession().dataTask(with: request) { _, response, error in
               DispatchQueue.main.async {
                    if let error = error {
                         call.reject("Failed to check folder existence: \(error.localizedDescription)")
                         return
                    }
                    
                    guard let httpResponse = response as? HTTPURLResponse else {
                         call.reject("Invalid response")
                         return
                    }
                    
                    if (200...299).contains(httpResponse.statusCode) {
                         call.resolve(["message": "Folder exists.", "exists": true])
                    } else {
                         call.reject("Folder does not exist: \(httpResponse.statusCode)")
                    }
               }
          }
          task.resume()
     }
     
     @objc func deleteFolder(_ call: CAPPluginCall) {
          guard let url = call.getString("url"),
                let username = call.getString("username"),
                let password = call.getString("password"),
                let requestURL = URL(string: url) else {
               call.reject("Missing or invalid parameters")
               return
          }
          
          var request = URLRequest(url: requestURL)
          request.httpMethod = "DELETE"
          request.addValue(getAuthHeader(username: username, password: password), forHTTPHeaderField: "Authorization")
          
          let task = getSession().dataTask(with: request) { _, response, error in
               DispatchQueue.main.async {
                    if let error = error {
                         call.reject("Delete failed: \(error.localizedDescription)")
                         return
                    }
                    
                    guard let httpResponse = response as? HTTPURLResponse else {
                         call.reject("Invalid response")
                         return
                    }
                    
                    if (200...299).contains(httpResponse.statusCode) {
                         call.resolve(["message": "Folder deleted"])
                    } else {
                         call.reject("Delete failed with status: \(httpResponse.statusCode)")
                    }
               }
          }
          task.resume()
     }
     
     @objc func uploadFile(_ call: CAPPluginCall) {
          guard let url = call.getString("url"),
                let username = call.getString("username"),
                let password = call.getString("password"),
                let base64Content = call.getString("content"),
                let data = Data(base64Encoded: base64Content),
                let requestURL = URL(string: url) else {
               call.reject("Missing or invalid parameters")
               return
          }
          
          var request = URLRequest(url: requestURL)
          request.httpMethod = "PUT"
          request.addValue(getAuthHeader(username: username, password: password), forHTTPHeaderField: "Authorization")
          request.httpBody = data
          
          let task = getSession().dataTask(with: request) { _, response, error in
               DispatchQueue.main.async {
                    if let error = error {
                         call.reject("Upload failed: \(error.localizedDescription)")
                         return
                    }
                    
                    guard let httpResponse = response as? HTTPURLResponse else {
                         call.reject("Invalid response")
                         return
                    }
                    
                    if (200...299).contains(httpResponse.statusCode) {
                         call.resolve(["message": "File uploaded"])
                    } else {
                         call.reject("Upload failed with status: \(httpResponse.statusCode)")
                    }
               }
          }
          task.resume()
     }
     
     @objc func getFile(_ call: CAPPluginCall) {
          guard let url = call.getString("url"),
                let username = call.getString("username"),
                let password = call.getString("password"),
                let requestURL = URL(string: url) else {
               call.reject("Missing or invalid parameters")
               return
          }
          
          var request = URLRequest(url: requestURL)
          request.httpMethod = "GET"
          request.addValue(getAuthHeader(username: username, password: password), forHTTPHeaderField: "Authorization")
          
          let task = getSession().dataTask(with: request) { data, response, error in
               DispatchQueue.main.async {
                    if let error = error {
                         call.reject("GET failed: \(error.localizedDescription)")
                         return
                    }
                    
                    guard let httpResponse = response as? HTTPURLResponse,
                          let data = data else {
                         call.reject("Invalid response")
                         return
                    }
                    
                    if (200...299).contains(httpResponse.statusCode) {
                         let base64Data = data.base64EncodedString()
                         call.resolve(["content": base64Data])
                    } else {
                         call.reject("GET failed with status: \(httpResponse.statusCode)")
                    }
               }
          }
          task.resume()
     }
     
     @objc func listContents(_ call: CAPPluginCall) {
          guard let url = call.getString("url"),
                let username = call.getString("username"),
                let password = call.getString("password"),
                let requestURL = URL(string: url) else {
               call.reject("Missing or invalid parameters")
               return
          }
          
          var request = URLRequest(url: requestURL)
          request.httpMethod = "PROPFIND"
          request.addValue(getAuthHeader(username: username, password: password), forHTTPHeaderField: "Authorization")
          
          // THIS IS THE KEY FIX - Add Depth header to get folder contents
          request.addValue("1", forHTTPHeaderField: "Depth")
          
          // Optional: Add Content-Type header
          request.addValue("application/xml", forHTTPHeaderField: "Content-Type")
          
          // Optional: Add a basic PROPFIND body to request specific properties
          let propfindBody = """
         <?xml version="1.0" encoding="utf-8" ?>
         <D:propfind xmlns:D="DAV:">
             <D:prop>
                 <D:resourcetype/>
                 <D:getcontentlength/>
                 <D:getlastmodified/>
                 <D:creationdate/>
             </D:prop>
         </D:propfind>
         """
          
          request.httpBody = propfindBody.data(using: .utf8)
          
          let task = getSession().dataTask(with: request) { data, response, error in
               DispatchQueue.main.async {
                    if let error = error {
                         call.reject("Failed to list folder contents: \(error.localizedDescription)")
                         return
                    }
                    
                    guard let httpResponse = response as? HTTPURLResponse else {
                         call.reject("Invalid response")
                         return
                    }
                    
                    if (200...299).contains(httpResponse.statusCode) {
                         let xmlResponse = data != nil ? String(data: data!, encoding: .utf8) ?? "No response body" : "No response body"
                         call.resolve([
                              "message": "Folder contents retrieved.",
                              "data": xmlResponse
                         ])
                    } else {
                         let responseString = data != nil ? String(data: data!, encoding: .utf8) ?? "No response body" : "No response body"
                         call.reject("Failed to list folder contents: \(httpResponse.statusCode) - \(responseString)")
                    }
               }
          }
          task.resume()
     }
}
